packages/astro/src/runtime/server/render/page.ts

```diff
@@ -85,6 +85,18 @@ export async function renderPage(
 	if (route?.component.endsWith('.md')) {
 		headers.set('Content-Type', 'text/html; charset=utf-8');
 	}
-	const response = new Response(body, { ...init, headers });
-	return response;
+	let status = init.status;
+	// Custom 404.astro and 500.astro are particular routes that must return a fixed status code
+	// The check uses the `/` to avoid possible false negatives such as custom_404.astro
+	if (route?.component.endsWith("/404.astro")) {
```

Comment:
What if the rewrite aims at `src/pages/foo/404.astro`? This would be also matched but shouln't right?

Reply:
I thought about it. I was thinking to pass `config.srcDir`, or check `pages/404.astro`. Which one should I go for?

Reply:
What are the valid values of `route.component`? It's doesn't include `src/pages` does it?

Reply:
They can differ, especially if the user changes the `srcDir` to something different from `src/`. However, it's safe to assume that these files are at the root for `pages/`, correct? If so, the value of `route.component` should always end with `pages/404.astro`

Reply:
Maybe not necessarily because of integrations?

Reply:
I am not sure what you mean, can you expand what you're talking about?

Reply:
I mean what if an integration does that?
```ts
injectRoute({
	pattern: '/404'
	entrypoint: './foo/Bar.astro'
})
```
Is `route.component` computed based on the entrypoint?

Reply:
@ematipico I don't think that's a safe assumption, you can have `src/pages/blog/pages/404.astro`, a `pages` folder as part of the URL.

Reply:
So what do guys suggest? Passing `srcDir` and concat it with `pages/404.astro`

Reply:
I updated the code to use `route` instead